### PR TITLE
Add new constructors and index template to team-level mdranges

### PR DESCRIPTION
**Pull request by tcclevenger**
Currently, `{TeamVector|TeamThread|ThreadVector}MDRange` only has a constructor for the upper iteration bound, and assumes 0 as the lower bound, and hard codes an index type `int`. Looking forward to self-similar interface, we need constructor which takes lower bound and policy traits can take in an index.

For testing, I changed team-level 4D mdranges tests to use the lower bound constructor.

### Changelog Entry
Expand team-level MDRanges to have non-trivial lower bound iterations and arbitrary index type.

### CODE REVIEW

File: core/src/impl/Kokkos_TeamMDPolicy.hpp [hunk 1]
```diff
@@ -103,26 +103,26 @@ KOKKOS_INLINE_FUNCTION auto nested_policy(
     TeamMDRangeMode<TeamMDRangeLastNestLevel::NotLastNestLevel,
                     TeamMDRangeParThread::ParThread,
                     TeamMDRangeParVector::NotParVector>,
-    TeamHandle const& team, int count) {
-  return TeamThreadRange(team, count);
+    TeamHandle const& team, int begin, int end) {
```
Comment (masterleinad) [inline core/src/impl/Kokkos_TeamMDPolicy.hpp]: Why is this still `int`?

File: core/src/Kokkos_ExecPolicy.hpp [hunk 2]
```diff
@@ -1000,21 +994,22 @@ KOKKOS_INLINE_FUNCTION void md_parallel_impl(TeamMDPolicy const& policy,
                                              ReductionValueType&& val);
 }  // namespace Impl
 
-template <typename Rank, typename TeamHandle>
+template <typename Rank, typename TeamHandle, typename iType = int>
```
Comment (masterleinad) [inline core/src/Kokkos_ExecPolicy.hpp]: Shouldn't the TeamHandle know about the index type?

File: core/src/impl/Kokkos_TeamMDPolicy.hpp (hunk 1 above)
Comment (tcclevenger) [reply core/src/impl/Kokkos_TeamMDPolicy.hpp]: Oversight. Fixed now.

File: core/src/Kokkos_ExecPolicy.hpp (hunk 2 above)
Comment (tcclevenger) [reply core/src/Kokkos_ExecPolicy.hpp]: I don't think TeamHandles have an index type?

Comment (masterleinad) [reply core/src/Kokkos_ExecPolicy.hpp]: Yeah, it makes sense for the policy to provide the type but we commonly use `Kokkos::IndexType<MyIndexType>` instead of bare `MyIndexType` template argument (which might mek CTAD for complicated).

Comment (tcclevenger) [reply core/src/Kokkos_ExecPolicy.hpp]: I modeled this after `TeamVectorRange` where we just pass in a type. For the MDRangePolicy self-sim stuff, like I did with RangePolicy, I'm planning on using index type from the traits which use `Kokkos::IndexType`

Comment (tcclevenger) [reply core/src/Kokkos_ExecPolicy.hpp]: Also ok to change it up for these boundary structs too though

**Review by crtrott — COMMENTED** [review]
Can we check what the impact of this is on register count for something like a rank 4 nested MDRange? We may wanna consider going a step further and having independent iType for lower an upper, and essentially allowing the lower to be `integral_constant<int, 0>`. We could get that via CTAD. This way you could still get compile time 0 lower bounds in the loops, and safe state in the class. Another step further would be to store (two different) Kokkos::extents as lower and upper and thus allowing for compile time bounds. I think we need to explore the compile and runtime impact of this. My gutfeeling is that it matters more here, than for the top level (from CPU launched) version. But eventually we would need to support it for both if we find it matters.

File: core/unit_test/TestTeamMDRange.hpp [hunk 3]
```diff
@@ -188,6 +200,11 @@ struct TestTeamThreadMDRangeParallelFor : public TestTeamMDParallelFor {
     int n1         = dims[2];
     int n2         = dims[3];
 
+    // For 4D, use lower bound constructors
+    int s0 = 1;
+    int s1 = 1;
+    int s2 = 1;
```
Comment (Adrien-Tab) [inline core/unit_test/TestTeamMDRange.hpp]: Maybe we could also test with negative indices ? Or `!= 1`

File: core/src/impl/Kokkos_TeamMDPolicy.hpp [hunk 4]
```diff
@@ -171,7 +171,8 @@ KOKKOS_INLINE_FUNCTION void nested_loop(
                                 next_nest_level>;
   using TeamMDNextMode = typename NextNestingTracker::RangeMode;
 
-  for (int i = 0; i != policy.boundaries[CurrentNestLevel]; ++i) {
+  for (int i = policy.lower[CurrentNestLevel];
```
[the remark this answers is not in the document]
Comment (tcclevenger) [reply core/src/impl/Kokkos_TeamMDPolicy.hpp]: I changed to using "auto"

File: core/src/impl/Kokkos_TeamMDPolicy.hpp [hunk 5]
```diff
@@ -206,7 +207,8 @@ KOKKOS_INLINE_FUNCTION void nested_loop(
   // This recursively processes ranks from [0..TotalNestLevel-1]
   // args... is passed by value because it should always be ints
   parallel_for(
-      nested_policy(mode, policy.team, policy.boundaries[CurrentNestLevel]),
+      nested_policy(mode, policy.team, policy.lower[CurrentNestLevel],
+                    policy.upper[CurrentNestLevel]),
       [&](int const& i) {
         if constexpr (Rank::outer_direction == Iterate::Right) {
```
[the remark this answers is not in the document]
Comment (tcclevenger) [reply core/src/impl/Kokkos_TeamMDPolicy.hpp:214]: I changed to using auto

File: core/unit_test/TestTeamMDRange.hpp (hunk 3 above)
Comment (ldh4) [reply core/unit_test/TestTeamMDRange.hpp]: Agree with this. We could use a test exercising non-uniform lower bounds.

File: core/src/Kokkos_ExecPolicy.hpp [hunk 6]
```diff
@@ -1102,26 +1130,35 @@ struct TeamVectorMDRange<Rank<N, OuterDir, InnerDir>, TeamHandle> {
...
-      : team(team_), boundaries{static_cast<BoundaryType>(args)...} {
-    static_assert(sizeof...(Args) == total_nest_level);
-  }
+      : team(team_), upper{static_cast<IndexType>(args)...} {}
+
+  // Constructor range between lower and upper bounds
+  KOKKOS_INLINE_FUNCTION TeamVectorMDRange(TeamHandleType const& team_,
+                                           const BoundaryType& lower_,
+                                           const BoundaryType& upper_)
+      : team(team_), lower(lower_), upper(upper_) {}
```
Comment (Adrien-Tab) [inline core/src/Kokkos_ExecPolicy.hpp]: Should we add new deduction guides for these constructors ?

**Comment by tcclevenger** [comment]
> Can we check what the impact of this is on register count for something like a rank 4 nested MDRange?

@crtrott I compiled a simple test program on V100 using `-Xptxas=-v`
```c++
Kokkos::parallel_for(
  TeamPolicy(1, auto), KOKKOS_LAMBDA(team) {
    Kokkos::parallel_for(TeamVectorMDRange(team, n0, n1, n2, n3), 
                         [=] (i,j,k,l){
      // Some FP computation/view write
    });
  });

Kokkos::parallel_for(
  TeamPolicy(1, auto), KOKKOS_LAMBDA(team) {
    Kokkos::parallel_for(TeamThreadMDRange(team, o0, o1, o2, o3), 
                         [=] (i,j,k,l){
      Kokkos::parallel_for(ThreadVectorMDRange(team, i0, i1, i2, i3), 
                           [=] (a,b,c,d){
        // Some FP computation/view write
      });
    });
  });
```
For both current develop (0 lower bound is compile time) and this version (storing runtime 0 lower bound) we get the exact same register usage
```
# Kernel with TeamThreadMDRange, TheadVectorMDRange
ptxas info    : Used 60 registers, 8 bytes smem, 576 bytes cmem[0], 8 bytes cmem[2]

# Kernel with TeamVectorMDRange
ptxas info    : Used 52 registers, 8 bytes smem, 560 bytes cmem[0], 8 bytes cmem[2]
```

Even still, it would not take much effort to have separate lower and upper types (as you suggest with std::integral_constant in the 0 lower bound case) and would align better with MDRangePolicy.

File: core/unit_test/TestTeamMDRange.hpp (hunk 3 above)
Comment (tcclevenger) [reply core/unit_test/TestTeamMDRange.hpp]: Agree, I'll add something
